chiron/mcmc.py

```diff
@@ -112,103 +170,802 @@ def run(
+        report_frequency: int = 1,
+        update_stepsize: bool = False,
+        update_stepsize_frequency: int = 100,
+        method: str = "metropolis",
     ) -> None:
-        super().__init__(nr_of_moves, reporter=reporter)
+        """
+        Initialize the move.
 
-    def apply_move(self):
+        Parameters
+        ----------
+        nr_of_moves
+            Number of moves to be applied in each call to update.
+        reporter
+            Reporter object for saving the simulation step data.
+        report_frequency
+            Frequency of saving the simulation data.
+        update_stepsize
+            Whether to update the "stepsize" of the move. Stepsize is a generic term for the key move parameters.
+            For example, for a simple displacement move this would be the displacement_sigma.
+        update_stepsize_frequency
+            Frequency of updating the stepsize of the move.
+        method
+            Methodology to use for accepting or rejecting the proposed state.
+            Default is "metropolis".
         """
-        Apply a Monte Carlo move to the system.
+        super().__init__(
+            nr_of_moves,
+            reporter=reporter,
+            report_frequency=report_frequency,
+        )
+        self.method = method  # I think we should pass a class/function instead of a string, like space.
 
-        This method should be overridden by subclasses to define specific types of moves.
+        self.reset_statistics()
+        self.update_stepsize = update_stepsize
+        self.update_stepsize_frequency = update_stepsize_frequency
 
-        Raises
-        ------
-        NotImplementedError
-            If the method is not implemented in subclasses.
+    def update(
+        self,
+        sampler_state: SamplerState,
+        thermodynamic_state: ThermodynamicState,
+        nbr_list: Optional[PairsBase] = None,
+    ) -> Tuple[SamplerState, ThermodynamicState, Optional[PairsBase]]:
         """
+        Perform the defined move and update the state.
 
-        raise NotImplementedError("apply_move() must be implemented in subclasses")
+        Parameters
+        ----------
+        sampler_state : SamplerState
+            The initial state of the simulation, including positions.
+        thermodynamic_state : ThermodynamicState
+            The thermodynamic state of the system, including temperature and potential.
+        nbr_list : PairBase, optional
+            Neighbor list for the system.
 
-    def compute_acceptance_probability(
+        Returns
+        -------
+        sampler_state : SamplerState
+            The updated sampler state.
+        thermodynamic_state : ThermodynamicState
+            The updated thermodynamic state.
+        nbr_list: PairsBase
+            The updated neighbor/pair list. If a nbr_list is not set, this will be None.
+        """
+        calculate_current_reduced_potential = True
+
+        for i in range(self.nr_of_moves):
```

Comment:
In some instances, we may want this loop to be efficiently `jit`ted too, but my expectation is that most of our time will be spent in Langevin dynamics instead. But it's something we may want to make a note of and ensure that we could optimize this if needed.